File: observer-example/observers.scala
class Person {
  def shout(text: String) { println(text) }
}

class Dog {
  def shout(text: String) { println(text) }
}

class ShoutyMonitor[T <: {def shout(text: String): Unit}] {
  var monitored: Option[T] = None
  
  def monitor(monitored: T) {
    this.monitored = Some[T](monitored)
  }
  
  def shout(text: String) {
    this.monitored.get.shout(text)
    println("LOG: " + text + " by a " + this.monitored.get.getClass.getName)
  }
}

/* Runtime: */
var sm = new ShoutyMonitor
sm.monitor(new Dog)
sm.shout("Hello")

File: observer-example/observers.rb
class Person
  def shout(text)
    puts text.to_s
  end
end

class Dog
  def shout(text)
    puts text.to_s
  end
end

class ShoutyMonitor
  def initialize
    @monitored = nil
  end
  
  def monitor(monitored)
    @monitored = monitored if monitored.responds_to?(:shout)
  end
  
  def shout(text)
    if @monitored.responds_to?(:shout)
      @monitored.shout(text.to_s) 
      println("LOG: " + text.to_s + " by a " + @monitored.class.to_s)
    end
  end
end

# Runtime:
sm = ShoutyMonitor.new
sm.monitor(Dog.new)
sm.shout("Hello")